tokio/src/runtime/task/trace/mod.rs

```diff
@@ -0,0 +1,246 @@
+            Context::with_current(|context_cell| {
+                if let Some(mut collector) = context_cell.collector.take() {
+                    let mut frames = vec![];
+                    let mut above_leaf = false;
+
+                    if let Some(active_frame) = context_cell.active_frame.get() {
+                        let active_frame = active_frame.as_ref();
+
+                        backtrace::trace(|frame| {
+                            let below_root =
+                                !ptr::eq(frame.symbol_address(), active_frame.inner_addr);
+
+                            // only capture frames above `Trace::leaf` and below
+                            // `Trace::root`.
+                            if above_leaf && below_root {
+                                frames.push(frame.to_owned().into());
+                            }
+
+                            if ptr::eq(frame.symbol_address(), Self::leaf as *const _) {
+                                above_leaf = true;
+                            }
+
+                            // only continue unwinding if we're below `Trace::root`
+                            below_root
+                        });
+                    }
+                    collector.backtraces.push(frames);
+                    context_cell.collector.set(Some(collector));
+                }
+            });
+        }
+    }
+}
+
+impl fmt::Display for Trace {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Tree::from_trace(self.clone()).fmt(f)
+    }
+}
+
+fn defer<F: FnOnce() -> R, R>(f: F) -> impl Drop {
+    use std::mem::ManuallyDrop;
+
+    struct Defer<F: FnOnce() -> R, R>(ManuallyDrop<F>);
+
+    impl<F: FnOnce() -> R, R> Drop for Defer<F, R> {
+        #[inline(always)]
+        fn drop(&mut self) {
+            unsafe {
+                ManuallyDrop::take(&mut self.0)();
+            }
+        }
+    }
+
+    Defer(ManuallyDrop::new(f))
+}
+
+impl<T: Future> Future for Root<T> {
+    type Output = T::Output;
+
+    #[inline(never)]
+    fn poll(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        // SAFETY: The context's current frame is restored to its original state
+        // before `frame` is dropped.
+        unsafe {
+            let mut frame = Frame {
+                inner_addr: Self::poll as *const c_void,
+                parent: None,
+            };
+
+            Context::with_current_frame(|current| {
+                frame.parent = current.take();
+                current.set(Some(NonNull::from(&frame)));
+            });
+
+            let _restore = defer(|| {
+                Context::with_current_frame(|current| {
+                    current.set(frame.parent);
+                });
+            });
+
+            let this = self.project();
+            this.future.poll(cx)
+        }
+    }
+}
+
+/// Trace and poll all tasks of the current_thread runtime.
+pub(in crate::runtime) fn trace_current_thread(
```

Comment:
Could we make the **trace_current_thread** function more readable and easier to understand? I think we should encapsulate some of the logic into separate functions to improve the function's clarity and maintainability.
Something like this (no tested)

```rust
/// Notify all tasks and return a vector of their raw pointers.
fn notify_all_tasks(owned: &OwnedTasks<Arc<current_thread::Handle>>) -> Vec<NonNull<task::RawTask>> {
    let mut tasks = vec![];
    owned.for_each(|task| {
        task.as_raw().state().transition_to_notified_for_tracing();
        tasks.push(task.as_raw());
    });
    tasks
}

/// Collect the traces for all tasks and return a vector of `Trace` objects.
fn collect_task_traces(tasks: Vec<NonNull<task::RawTask>>) -> Vec<Trace> {
    tasks
        .into_iter()
        .map(|task| {
            let ((), trace) = Trace::capture(|| task.poll());
            trace
        })
        .collect()
}

/// Trace and poll all tasks of the current_thread runtime.
pub(in crate::runtime) fn trace_current_thread(
    owned: &OwnedTasks<Arc<current_thread::Handle>>,
    local: &mut VecDeque<Notified<Arc<current_thread::Handle>>>,
    injection: &mut VecDeque<Notified<Arc<current_thread::Handle>>>,
) -> Vec<Trace> {
    local.clear();
    injection.clear();
    let tasks = notify_all_tasks(owned);
    collect_task_traces(tasks)
}
```